File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class FeatureDevelopment {

    private static final int MAX_PROCESS = 100;

    private final Queue<Integer> queue;

    public FeatureDevelopment() {
        queue = new LinkedList<>();
    }
    
    public int calculateRemainDate(final int progress, final int speed) {
        double remain = (100 - progress) / (double) speed;
        return (int) Math.ceil(remain);
    }

    public int[] sol(int[] progresses, int[] speeds) {
        List<Integer> dates = new ArrayList<>();

        for (int i = 0; i < progresses.length; i++) {
            int remainDate = calculateRemainDate(progresses[i], speeds[i]);

            if (!queue.isEmpty() && queue.peek() < remainDate) {
                dates.add(queue.size());
                queue.clear();
            }
            queue.offer(remainDate);
        }
        dates.add(queue.size());
        queue.clear();

        return dates.stream()
            .mapToInt(Integer::intValue)
            .toArray();
    }
}


class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        final Queue<Integer> days = new LinkedList<>();
        final List<Integer> deploys = new ArrayList<>();

        for (int i = 0; i < progresses.length; i++) {
            int remainDays = calculateRemainDays(progresses[i], speeds[i]);

            if (!days.isEmpty() && days.peek() < remainDays) {
                deploys.add(days.size());
                days.clear();
            }
            days.offer(remainDays);
        }

        if (!days.isEmpty()) {
            deploys.add(days.size());
            days.clear();
        }
        return deploys.stream()
                .mapToInt(Integer::intValue)
                .toArray();
    }

    private int calculateRemainDays(final int progress, final int speed) {
        return (int) Math.ceil((100.0 - progress) / speed);
    }

}

File: 프로그래머스/2/42586. 기능개발/기능개발.py
"""
- 현재 작업에 대해 배포까지 남은 기간을 계산한다.
    - 93 -> 7
- 다음 작업의 배포 기간을 계산한다.
    - 30 -> 3 
- if 현재 작업 <= 다음 작업 -> 스택에 넣음
- else -> pop & 다음 작업을 스택에 넣음
"""
def rest_time(progress, speed):
    from math import ceil
    return ceil((100 - progress) / speed)
    
def solution(progresses, speeds):
    answer = []
    stk = []
    
    for progress, speed in zip(progresses, speeds):
        curr = rest_time(progress, speed)

        if stk and stk[0] < curr:
            answer.append(len(stk))
            stk.clear()
            
        stk.append(curr)

    if stk:
        answer.append(len(stk))
    
    return answer